src/TelemetryChannels/ServerTelemetryChannel/Shared/AdaptiveSamplingTelemetryProcessor.cs

```diff
@@ -0,0 +1,265 @@
+﻿namespace Microsoft.ApplicationInsights.WindowsServer.TelemetryChannel
+{
+    using System;
+
+    using Microsoft.ApplicationInsights.Channel;
+    using Microsoft.ApplicationInsights.Extensibility;
+    using Microsoft.ApplicationInsights.WindowsServer.Channel.Implementation;
+
+    /// <summary>
+    /// Telemetry processor for sampling telemetry at a dynamic rate before sending to Application Insights.
+    /// </summary>
+    public class AdaptiveSamplingTelemetryProcessor : ITelemetryProcessor, IDisposable
+    {
+        /// <summary>
+        /// Fixed-rate sampling telemetry processor.
+        /// </summary>
+        private SamplingTelemetryProcessor samplingProcessor;
+
+        /// <summary>
+        /// Sampling percentage estimator telemetry processor.
+        /// </summary>
+        private SamplingPercentageEstimatorTelemetryProcessor estimatorProcessor;
+
+        /// <summary>
+        /// Sampling percentage estimator settings.
+        /// </summary>
+        private SamplingPercentageEstimatorSettings estimatorSettings;
+
+        /// <summary>
+        /// Initial sampling percentage applied after start of the process.
+        /// </summary>
+        private double initialSamplingPercentage;
+
+        /// <summary>
+        /// Callback invoked every time sampling percentage is evaluated.
+        /// </summary>
+        private AdaptiveSamplingPercentageEvaluatedCallback evaluationCallback;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AdaptiveSamplingTelemetryProcessor"/> class.
+        /// <param name="next">Next TelemetryProcessor in call chain.</param>
+        /// </summary>
+        public AdaptiveSamplingTelemetryProcessor(ITelemetryProcessor next)
+            : this(new SamplingPercentageEstimatorSettings(), null, next)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AdaptiveSamplingTelemetryProcessor"/> class.
+        /// <param name="settings">Sampling percentage estimator settings.</param>
+        /// <param name="callback">Callback invoked every time sampling percentage is evaluated.</param>
+        /// <param name="next">Next TelemetryProcessor in call chain.</param>
+        /// </summary>
+        public AdaptiveSamplingTelemetryProcessor(
+            SamplingPercentageEstimatorSettings settings,
+            AdaptiveSamplingPercentageEvaluatedCallback callback,
+            ITelemetryProcessor next)
+        {
+            this.estimatorSettings = settings;
+            this.evaluationCallback = callback;
+
+            // make estimatortelemetry processor  work after sampling was done
+            this.estimatorProcessor = new SamplingPercentageEstimatorTelemetryProcessor(settings, this.SamplingPercentageChanged, next);
+            this.samplingProcessor = new SamplingTelemetryProcessor(this.estimatorProcessor);
```

Comment:
Let's use DI to inject the sampling processor here. As you mentioned that one of the reason to have two sampling processors - one simple and one Adaptive is to be able to replace implementation of sampling rate calculation, but keep "adaptive" part of it intact. Maybe ```SaplingTelemetryProcessor``` should sign up on sampling interval change callback or something...